File: Project Hippopotamus/Content/Lua/Terrain Processors/TerrainWorldProcessor.lua
-- FIXME
-- terrainValleyProcessor = require("TerrainValleyProcessor") 

worldData = {}
worldWidth = 0
worldHeight = 0

function BeginProcess(width, height)
    worldWidth = width
    worldHeight = height
    Generate()
    -- terrainValleyProcessor.Generate()

    Logger.Log("TerrainWorldProcessor", "I WORK!")
end

function Lerp(a, b, time)
    return a + (b - a) * time
end

function Generate()
    local heightMap = {}
    groundHeight = worldHeight * 0.7

    -- { MinimumAmplitude, MaximumAmplitude, MinimumFrequency, MaximumFrequency }
    curveParameters = 
    {
        { 0, 0.3, 0.5, 1.5 }, 
        { 0, 0.01, 5.0, 12.0 },
        { 0, 0.01, 5.0, 12.0 }
    }

    noiseChance = 0.05
    noiseMinimumMagnitude = -2
    noiseMaxMagnitude = 1

    for x = 0, worldWidth do
        heightMap[x] = groundHeight
    end

    for _, curveParameter in pairs(curveParameters) do
        amplitude = worldHeight * Lerp(curveParameter[1], curveParameter[2], Random.Value())
        frequency = Lerp(curveParameter[3], curveParameter[4], Random.Value()) / 100

        offset = 0
        phase = Random.Value() * worldWidth
        for x = 0, worldWidth do
            heightMap[x] = heightMap[x] + amplitude * math.sin(frequency * x - phase) + offset
        end
    end

    -- Do noise!
    for x = 0, worldWidth do
        if Random.Value() < noiseChance then
            heightMap[x] = heightMap[x] + Lerp(noiseMinimumMagnitude, noiseMaxMagnitude, Random.Value())
        end
    end

    for x = 0, worldWidth do
        for y = 0, worldHeight do
            if worldHeight - 1 - y <= heightMap[x] then
                worldData[x * worldHeight + y] = true--TileType.Grass
            end
        end
    end
end

File: Assets/Game/Scripts/World/TerrainWorldProcessor.cs
using UnityEngine;

public class TerrainWorldProcessor : ITerrainProcessor
{
    public void Generate(WorldData worldData)
    {
        double[] heightMap = new double[worldData.Width];
        double groundHeight = worldData.Height * 0.7f;
        SineCurveParameter[] curveParameters =
        {
              new SineCurveParameter(0.0, 0.3, 0.5, 1.5),     // Main big terrain feature
              new SineCurveParameter(0.0, 0.01, 5.0, 12.0),   // Medium scale randomization
              new SineCurveParameter(0.0, 0.01, 5.0, 12.0),  // Medium scale randomization
        };

        // Frequency of the noise
        const double noiseChance = 0.05;
        const double noiseMinimumMagnitude = -2.0;
        const double noiseMaxMagnitude = 1.0;

        for (int x = 0; x < worldData.Width; x++)
        {
            heightMap[x] = groundHeight;
        }

        foreach (SineCurveParameter curveParameter in curveParameters)
        {
            double amplitude = worldData.Height * Mathf.Lerp((float)curveParameter.MinimumAmplitude, (float)curveParameter.MaximumAmplitude, Random.value);
            double frequency = Mathf.Lerp((float)curveParameter.MinimumFrequency, (float)curveParameter.MaximumFrequency, Random.value) / 100.0;

            const double offset = 0.0;
            double phase = Random.value * worldData.Width;
            for (int x = 0; x < worldData.Width; x++)
            {
                heightMap[x] += amplitude * System.Math.Sin(frequency * x - phase) + offset;
            }
        }

        for (int x = 0; x < worldData.Width; x++)
        {
            if (Random.value < noiseChance)
            {
                heightMap[x] += Mathf.Lerp((float)noiseMinimumMagnitude, (float)noiseMaxMagnitude, Random.value);
            }
        }

        for (int x = 0; x < worldData.Width; x++)
        {
            for (int y = 0; y < worldData.Height; y++)
            {
                if (y <= heightMap[x])
                {
                    worldData.Tiles[x, y] = TileType.Grass;
                }
            }
        }
    }

}
